source/retryable-reads/tests/README.rst

```diff
@@ -0,0 +1,269 @@
+
+  - ``description``: The name of the test.
+
+  - ``clientOptions``: Optional, parameters to pass to MongoClient().
+
+  - ``failPoint``: Optional, a server failpoint to enable expressed as the
+    configureFailPoint command to run on the admin database.
+
+  - ``sessionOptions``: Optional, parameters to pass to
+    MongoClient.startSession().
+
+  - ``operation``: A document describing an operation to be
+    executed. The document has the following fields:
+
+    - ``name``: The name of the operation on ``object``.
+
+    - ``object``: The name of the object to perform the operation on. Can be
+      "database", "collection", or "client."
+
+    - ``collectionOptions``: Optional, parameters to pass to the Collection()
+      used for this operation.
+
+    - ``command_name``: Present only when ``name`` is "runCommand". The name
+      of the command to run. Required for languages that are unable preserve
+      the order keys in the "command" argument when parsing JSON/YAML.
+
+    - ``arguments``: Optional, the names and values of arguments.
+
+    - ``result``: The return value from the operation, if any. This field may
+      be a scalar (e.g. in the case of a count), a single document, or an array
+      of documents in the case of a multi-document read. If the operation is
+      expected to return an error, the ``result`` is a single document that has
+      one or more of the following fields:
+
+      - ``error``: Optional. If ``true``, the test should expect an error or
+        exception. Implicitly true if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or
+        ``errorLabelsOmit``. If ``false``, the test should expect no error or
+        exception, even if ``result`` contains ``errorContains``,
+        ``errorCodeName``, ``errorLabelsContain``, and/or ``errorLabelsOmit``.
+
+      - ``errorContains``: Optional. A substring of the expected error message.
+
+      - ``errorCodeName``: Optional. The expected "codeName" field in the server
+        error response.
+
+      - ``errorLabelsContain``: Optional. A list of error label strings that the
+        error is expected to have.
+
+      - ``errorLabelsOmit``: Optional. A list of error label strings that the
+        error is expected not to have.
+
+  - ``expectations``: Optional list of command-started events.
+
+MapReduce Tests
+===============
+
+Drivers MUST ensure that ``mapReduce`` is retried only when ``out: { inline:
+1}``.  The tests for ``mapReduce`` should test against the same conditions as
+described in ``aggregate.yml`` and ``aggregate-serverErrors.yml``, with the
+exception that ``mapReduce`` MUST not retry when the `out` stage is not
+``{inline: 1}``.
+
+A recommended data set would be three documents that contain values for x such
+that ``x:0``, ``x:1`` and ``x:2``. The map function should increment each of
+these values by one, and the reduce function should add all x's together.
```

Comment:
If this is a prose test, I don't know if it's helpful to refer to YAML tests. Since those can be synced after the fact, it makes it all the more likely that drivers' mapReduce prose tests may become inconsistent (even within the driver itself). I think a single `mapReduce` command document, which drivers can translate for their mapReduce helper (if one exists) would be sufficient. You can advise them how to setup the fail point here and assert that the command is successful. Repeating the same test without `{inline: 1}` could then expect a failure.

Reply:
After reading "Optionally Retryable Commands" below, I understand why you referred to YAML tests here. Those likely contain the varieties of failures to test. If this is common to both the mapReduce and "other optional retryable commands", then perhaps you can just summarize those in this test README. I'm not sure if that makes it any more or less likely that these prose tests could get out of sync. I don't feel strongly about this, so if you believe referring to YAML is best I trust your judgement.

Reply:
Thanks Jeremy: I think I'll continue referring to the YAML :)